test/webgpu_required_limits.c

```diff
@@ -46,43 +43,50 @@ void assertLimitsCompatible(WGPULimits required_limits,
 #undef ASSERT_LIMITS_COMPATIBLE
 }
 
-void on_device_request_ended(WGPURequestDeviceStatus status,
-                             WGPUDevice device,
-                             char const* message,
-                             void* userdata) {
+static void on_device_request_ended(WGPURequestDeviceStatus status,
+                                    WGPUDevice device,
+                                    WGPUStringView message,
+                                    void* userdata1, void* userdata2) {
   assert(status == WGPURequestDeviceStatus_Success);
 
-  WGPUSupportedLimits device_supported_limits;
+  WGPULimits device_supported_limits = {0};
   wgpuDeviceGetLimits(device, &device_supported_limits);
 
   // verify that the obtained device fullfils required limits 
-  assertLimitsCompatible(adapter_supported_limits.limits,
-                         device_supported_limits.limits);
+  assertLimitsCompatible(adapter_supported_limits,
+                         device_supported_limits);
+
+  wgpuDeviceRelease(device);
+  exit(0);
 }
 
-void on_adapter_request_ended(WGPURequestAdapterStatus status,
-                              WGPUAdapter adapter,
-                              char const* message,
-                              void* userdata) {
+static void on_adapter_request_ended(WGPURequestAdapterStatus status,
+                                     WGPUAdapter adapter,
+                                     WGPUStringView message,
+                                     void* userdata1, void* userdata2) {
   assert(status == WGPURequestAdapterStatus_Success);
 
   wgpuAdapterGetLimits(adapter, &adapter_supported_limits);
 
-  // for device limits, require the limits supported by adapter
-  WGPURequiredLimits device_required_limits = {0,};
-  device_required_limits.limits = adapter_supported_limits.limits;
+  WGPUDeviceDescriptor device_desc = {
+    // for device limits, require the limits supported by adapter
+    .requiredLimits = &adapter_supported_limits,
+  };
+  wgpuAdapterRequestDevice(adapter, &device_desc, (WGPURequestDeviceCallbackInfo){
+    .mode = WGPUCallbackMode_AllowSpontaneous,
+    .callback = on_device_request_ended,
+  });
 
-  WGPUDeviceDescriptor device_desc = {0,};
-  device_desc.requiredFeatureCount = 0;
-  device_desc.requiredLimits = &device_required_limits;
-  wgpuAdapterRequestDevice(adapter, &device_desc, on_device_request_ended, NULL);
+  wgpuAdapterRelease(adapter);
 }
 
-int main() {
+int main(void) {
```

Comment:
Are these test changes coming from upstream or is emscripten (here) the source of truth for them?

(Wondering where this 'void' comes from because I don't think it should be needed by the compiler)

Reply:
`int main(void) {}` in a C function definition does nothing, it is the same as `int main() {}`.

Only in C declarations does `int main();` mean `int main(...);`, but not in definitions.

Reply:
Ah, I enabled `-Weverything` at some point because I was trying to track down a bug in some of these tests. No longer needed.